### ISSUE TITLE
Assembly results problem

### ISSUE THREAD

**Issue by wshuai294** [DaehwanKimLab/hisat-genotype#47]
Dear Chris,

I have simulated some samples to test Hisat-genotype's assembly performance.  I have found it works perfectly for some genes like HLA-B and HLA-C. However, I found it did not work well with HLA-A gene. Moreover, sometimes the sequences of the two haplotypes are the same while the ground truth is not. 

Maybe I ran it in the wrong way. It is also possible that I interpret the output results in fault. So could you please tell me how to interpret the output *report file? In case I worry that I make mistakes evaluating the Hisat-genotype. I will be very pleased if you afford your evaluation results about the assembly to me for reference.

In addition, sometimes when I ran it on HLA-DRB1 gene, it showed:
"
Traceback (most recent call last):
  File "/home/wangmengyao/packages/hisat-genotype-top/hisatgenotype_locus.py", line 2630, in <module>
    debug)
  File "/home/wangmengyao/packages/hisat-genotype-top/hisatgenotype_locus.py", line 2365, in genotyping_locus
    assembly_verbose)
  File "/home/wangmengyao/packages/hisat-genotype-top/hisatgenotype_locus.py", line 1241, in typing
    if read_seq[read_pos] != ref_seq[pos]:
IndexError: string index out of range
". Could you help me to figure it out?

Best wishes,
WANG Shuai

**Comment by chbe-helix** [comment]
Hi Wang Shuai,

There are some known bugs in the assembly code that are on the list to fix. To know if this is caused by one of them, could you let me know what versions of HISATgenotype, HISAT2, and how you ran the scripts? 

Thanks,
Chirs

**Comment by wshuai294** [comment]
Hi Chris,

The versions are here:
HISAT2 version 2.2.0 by Daehwan Kim
hisat-genotype-top version 2.2.0

The run script is:
python /home/wangmengyao/packages/hisat-genotype-top/hisatgenotype_locus.py --base hla --locus-list DRB1  --assembly -1 /mnt/disk2_workspace/wangmengyao/NeedleHLA/simu_data/novel/simu/fastq/HLA_12_T_50-50/HLA_12_T_50-50.read1.fastq.gz -2 /mnt/disk2_workspace/wangmengyao/NeedleHLA/simu_data/novel/simu/fastq/HLA_12_T_50-50/HLA_12_T_50-50.read2.fastq.gz --output-base output/HLA_12_T_50-50  --read-len 150

Thanks for your attention.
WANG Shuai

**Comment by chbe-helix** [comment]
Hi WANG Shuai,

It looks like you may be using an older version of HISAT-genotype. Can you confirm for me that you got HISAT-genotype in accordance with the manual here:

https://daehwankimlab.github.io/hisat-genotype/manual/

Thanks,
Chris

**Comment by wshuai294** [comment]
Hi Chris,

I follow the manual to install a new version of HISAT-genotype, the version is HISAT-genotype 1.3.2.  However I got a new bug:
"Script exited with error 1"
when I ran
/home/wangshuai/softwares/hisatgenotype/hisatgenotype_toolkit locus  --ref-genome /home/wangmengyao/packages/hisat-genotype-top/HLA_typing/genotype_genome --base hla --locus-list  A,B,C,DPA1,DPB1,DQA1,DQB1,DRB1 --assembly -1 /mnt/disk2_workspace/wangmengyao/NeedleHLA/simu_data/AlleleImbalance/d_5_5/simu/fastq/HLA_1_T_50-50/HLA_1_T_50-50.read1.fastq.gz -2 /mnt/disk2_workspace/wangmengyao/NeedleHLA/simu_data/AlleleImbalance/d_5_5/simu/fastq/HLA_1_T_50-50/HLA_1_T_50-50.read2.fastq.gz --read-len 150 -p 4 --out-dir ./ --assembly-name HLA_1_T_50-50

Can you help me with this?

Thanks,
WANG Shuai

**Comment by chbe-helix** [comment]
Hi WANG,

Try running the script without the `--assembly` options. If it doesn't run, there may be an incompatible version of a dependency installed. The most common dependency errors are python version (need >3 ideally >3.7) or samtools (need >1.3). Let me know what you find.

Thanks,
Chris

**Comment by wshuai294** [comment]
Hi Chris,

Now I can run HISAT-genotype successfully. And the typing result is perfect. However, I still have a problem. After I checked the assembly results, I sometimes found the output fasta sequence can not blast to the ground truth.
For example:
Database: HLA_9_T_50-50/DRB1.simu.fa
           1 sequences; 11,229 total letters



Query= DRB1.1

Length=18420


***** No hits found *****

Sometimes the output sequence can be mapped to the ground truth well, which means the run script is correct. So I don't know the reason why sometimes the output sequence is totally different from the truth. Could you please help me?

Best,
WANG Shuai

**Comment by chbe-helix** [comment]
Hi WANG,

That's an interesting issue I have not encountered yet. Would you be willing to send me the HISATgenotype results, the ground truth, and how you ran blast? I am having trouble visualizing exactly what is happening and what you are seeing from your example.

Thanks,
Chris

**Comment by wshuai294** [comment]
Hi, Chris,

I uploaded the simulation data and results to the link:
https://drive.google.com/file/d/1l5uh6RSoP8O-BeoSwdXt22kI3do8QTU8/view?usp=sharing

Please kindly find the file, In the folder.

The simulated fasta files (ground truth): HLA_9.simu.all.hla.ref; the novel allele means we added 10 mutations to the known allele.
The simulated fastq files: HLA_9_T_50-50.read*.fastq
The results of hisat2-genotype: assembly_graph-hla.HLA_9_T_50-50_read1_fastq_gz*


And the blast script is writen in perl like:
`/home/wangmengyao/packages/ncbi-blast-2.9.0+/bin/makeblastdb -in $sample/$hla.simu.fa -dbtype nucl -parse_seqids -out $sample/$hla.simu`;
`/home/wangmengyao/packages/ncbi-blast-2.9.0+/bin/blastn -query $sample/$hla.1.fa -out $sample/$hla.allele1.blast.xml -db $sample/$hla.simu -outfmt 5 -penalty -1 -reward 1 -gapopen 4 -gapextend 1 -strand plus`;


Thanks.

Best,
WANG Shuai

**Comment by chbe-helix** [comment]
Hi WANG,

Thanks for sending the information! Using the web-based BLAST I was able to get the novel DRB1 allele to align to the assembled sequence. I did notice that my output contains non-standard nucleotide symbols, most notably "D" to represent deletions. It may be a good idea to remove these nucleotides before trying your blast program. I'll edit the code to remove these deletion markers in the next HISATgenotype release. Let me know if this helps!

Thanks,
Chris
